File: common-utils/src/main/java/io/github/lagom130/lab/common/utils/enums/utils/EnumUtils.java
package io.github.lagom130.lab.common.utils.enums.utils;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.Objects;
import java.util.Optional;

/**
 * @author lujc
 * @date 2023/9/11.
 */
public class EnumUtils {

    /**
     * 忽略大小写转换枚举
     * 转换字段需要注解@EnumKey
     * 没有对应枚举返回null
     * @param value
     * @param enumClass
     * @return
     * @param <E>
     */
    public static <E extends Enum<E>> E parseIgnoreCase(String value, Class<E> enumClass) {
        if(Objects.isNull(value)) {
            return null;
        }
        if(Objects.isNull(enumClass)) {
            throw new IllegalArgumentException("enumClass cannot be null");
        }
        Field field = findEnumKeyFiled(enumClass)
                .orElseThrow(() -> new IllegalArgumentException(
                        enumClass + " must have a filed with the annotation " + EnumKey.class));
        field.setAccessible(Boolean.TRUE);
        return Arrays.stream(enumClass.getEnumConstants())
                .filter(enumConstant -> {
                    try {
                        return value.equalsIgnoreCase(String.valueOf(field.get(enumConstant)));
                    } catch (IllegalAccessException e) {
                        throw new RuntimeException(e);
                    }
                })
                .findFirst()
                .orElse(null);
    }

    private static <E extends Enum<E>> Optional<Field> findEnumKeyFiled(Class<E> enumClass) {

        return Arrays.stream(enumClass.getDeclaredFields())
                .filter(field -> Objects.nonNull(field.getDeclaredAnnotation(EnumKey.class)))
                .findFirst();

    }
}

File: zone-study-java/delivery-bird-service/src/main/kotlin/io/github/lagom130/lab/util/EnumUtils.kt
package io.github.lagom130.lab.util

import com.baomidou.mybatisplus.annotation.EnumValue
import java.lang.Boolean
import java.lang.reflect.Field
import java.util.*
import kotlin.Enum
import kotlin.IllegalArgumentException
import kotlin.RuntimeException
import kotlin.require

/**
 * @author lujc
 * @date 2023/11/2.
 */
object EnumUtils {
    /**
     * 转换字段需要注解@com.baomidou.mybatisplus.annotation.EnumValue
     * 没有对应枚举返回null
     * @param value
     * @param enumClass
     * @return
     * @param <E>
    </E> */
    fun <E : Enum<E>?, T> parse(value: T, enumClass: Class<E>): E {
        require(!Objects.isNull(enumClass)) { "enumClass cannot be null" }
        val field = findEnumKeyFiled(enumClass)
                .orElseThrow {
                    IllegalArgumentException(
                            enumClass.toString() + " must have a filed with the annotation " + EnumValue::class.java)
                }
        field.setAccessible(Boolean.TRUE)
        return Arrays.stream(enumClass.getEnumConstants())
                .filter { enumConstant: E? ->
                    try {
                        return@filter field[enumConstant] == value
                    } catch (e: IllegalAccessException) {
                        throw RuntimeException(e)
                    }
                }
                .findFirst()
                .orElse(null)
    }

    private fun <E : Enum<E>?> findEnumKeyFiled(enumClass: Class<E>): Optional<Field> {
        return Arrays.stream(enumClass.getDeclaredFields())
                .filter { field: Field -> Objects.nonNull(field.getDeclaredAnnotation(EnumValue::class.java)) }
                .findFirst()
    }
}
